geomesa-stream/geomesa-stream-datastore/src/main/scala/org/locationtech/geomesa/stream/datastore/StreamDataStore.scala

```diff
@@ -0,0 +1,214 @@
+import org.opengis.filter.spatial.{BBOX, BinarySpatialOperator, Within}
+import org.opengis.filter.{And, Filter, IncludeFilter, Or}
+
+import scala.collection.JavaConversions._
+
+case class FeatureHolder(sf: SimpleFeature, env: Envelope) {
+  override def hashCode(): Int = sf.hashCode()
+
+  override def equals(obj: scala.Any): Boolean = obj match {
+    case other: FeatureHolder => sf.equals(other.sf)
+    case _ => false
+  }
+}
+
+trait StreamListener {
+  def onNext(sf: SimpleFeature): Unit
+}
+
+class StreamDataStore(source: SimpleFeatureStreamSource, timeout: Int) extends ContentDataStore {
+  
+  val sft = source.sft
+  source.init()
+  val qt = new SynchronizedQuadtree
+
+  val cb =
+    CacheBuilder
+      .newBuilder()
+      .expireAfterWrite(timeout, TimeUnit.SECONDS)
+      .removalListener(
+        new RemovalListener[String, FeatureHolder] {
+          def onRemoval(removal: RemovalNotification[String, FeatureHolder]) = {
+            qt.remove(removal.getValue.env, removal.getValue.sf)
+          }
+        }
+      )
+
+  val features: Cache[String, FeatureHolder] = cb.build()
+
+  val listeners = Collections.synchronizedList(Lists.newArrayList[StreamListener]())
+
+  Executors.newSingleThreadExecutor().submit(
+    new Runnable {
+      override def run(): Unit = {
+        while(true) {
+          try {
+            val sf = source.next
+            if(sf != null) {
+              val env = sf.geometry.getEnvelopeInternal
+              qt.insert(env, sf)
+              features.put(sf.getID, FeatureHolder(sf, env))
+              listeners.foreach { l =>
+                try {
+                  l.onNext(sf)
+                } catch {
+                  case t: Throwable => getLogger.log(Level.WARNING, "Unable to notify listener", t)
+                }
+              }
+            }
+          } catch {
+            case t: Throwable =>
+            // swallow
+          }
+        }
+      }
+    }
+  )
+
+  override def createFeatureSource(entry: ContentEntry): ContentFeatureSource =
+    new StreamFeatureStore(entry, null, features, qt, sft)
+
+  def registerListener(listener: StreamListener): Unit = listeners.add(listener)
+
+  override def createTypeNames(): ju.List[Name] = Lists.newArrayList(sft.getName) 
+}
+
+class StreamFeatureStore(entry: ContentEntry,
+                         query: Query,
+                         features: Cache[String, FeatureHolder],
+                         qt: SynchronizedQuadtree,
+                         sft: SimpleFeatureType)
+  extends ContentFeatureStore(entry, query) {
+
+  type FR = FeatureReader[SimpleFeatureType, SimpleFeature]
+  type DFR = DelegateFeatureReader[SimpleFeatureType, SimpleFeature]
+  type DFI = DelegateFeatureIterator[SimpleFeature]
+
+
+  override def canFilter: Boolean = true
+
+  override def getBoundsInternal(query: Query) =
+    ReferencedEnvelope.create(new Envelope(-180, 180, -90, 90), DefaultGeographicCRS.WGS84)
```

Comment:
Should use org.locationtech.geomesa.utils.geotools.package$#CRS_EPSG_4326 - that way it will use the full crs if available (e.g. in geoserver).